src/datepicker/DatePickerState.ts

```diff
@@ -1,53 +1,105 @@
-import React from "react";
-import { useCompositeState, useDisclosureState } from "reakit";
+/**
+ * All credit goes to [React Spectrum](https://github.com/adobe/react-spectrum)
+ * We improved the Calendar from Stately [useDatePickerState](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/datepicker/src/useDatePickerState.ts)
+ * to work with Reakit System
+ */
 
-import { useDatePickerFieldState } from "./DatePickerFieldState";
-import { useCalendarState, CalendarStateInitialProps } from "../calendar";
+import * as React from "react";
+import { useControllableState } from "@chakra-ui/hooks";
+import {
+  useCompositeState,
+  usePopoverState,
+  unstable_useId as useId,
+} from "reakit";
 
-export interface DatePickerStateInitialProps extends CalendarStateInitialProps {
-  visible?: boolean;
-  initialDate?: Date;
-  dateFormat?: Intl.DateTimeFormatOptions;
-}
+import { useCalendarState } from "../calendar";
+import { setTime, isInvalid } from "./__utils";
+import { useDatePickerFieldState } from "./DatePickerFieldState";
+import { DatePickerStateInitialProps, ValidationState } from "./index.d";
 
 export const useDatePickerState = (props: DatePickerStateInitialProps = {}) => {
-  const { visible, dateFormat, initialDate = new Date() } = props;
+  const {
+    value: initialValue,
+    defaultValue = new Date(),
+    onChange,
+    minValue,
+    maxValue,
+    isDisabled,
+    isReadOnly,
+    isRequired,
+    pickerId: pickerIdProp,
+    dialogId: dialogIdProp,
+    formatOptions,
+    placeholderDate,
+  } = props;
 
-  const [date, setDate] = React.useState(initialDate);
+  const { id: pickerId } = useId({ id: pickerIdProp, baseId: "picker" });
+  const { id: dialogId } = useId({ id: dialogIdProp, baseId: "dialog" });
 
-  const segmentComposite = useCompositeState({ orientation: "horizontal" });
-  const disclosure = useDisclosureState({ visible });
-
-  const calendar = useCalendarState({
-    value: date,
-    onChange: date => {
-      setDate(new Date(date));
-      disclosure.hide();
-    },
+  const [value, setValue] = useControllableState({
+    value: initialValue,
+    defaultValue,
+    onChange,
+    shouldUpdate: (prev, next) => prev !== next,
   });
+  const dateValue = value != null ? new Date(value) : undefined;
 
+  // Intercept setValue to make sure the Time section is not changed by date selection in Calendar
+  const selectDate = (newValue: Date) => {
+    if (dateValue) {
+      setTime(newValue, dateValue);
+    }
```

Comment:
Fix the type by accepting DateValue instead of Date

```js
  const selectDate = (newValue: DateValue) => {
    if (dateValue) {
      setTime(new Date(newValue), dateValue);
    }

    setValue(newValue);
    popover.hide();
  };
```
And remove the @ts-ignore at line 69.